### PR TITLE
Add ComponentAuthProvider and AuthNClient, update AuthProvider.

### PR DESCRIPTION
**Pull request by ggallien13**

### CODE REVIEW

File: packages/react-auth/src/AuthNClient/index.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    this.config = {
+      ...config,
+    };
+  }
+
+  /*
+    General AuthN functions
+  */
+
+  async logout(userToken?: string): Promise<ClientResponse> {
```
Comment (cj-f) [inline packages/react-auth/src/AuthNClient/index.ts:30]: no "?"

File: packages/react-auth/src/AuthNFlow/client.ts [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+  FlowStart,
+  FlowSignupPassword,
+  FlowVerifyPassword,
+  FlowVerifyCallback,
+  FlowVerifyCaptcha,
+  FlowMfaStart,
+  FlowMfaComplete,
+} from "./types";
+
+interface ClientResponse {
```
Comment (cj-f) [inline packages/react-auth/src/AuthNFlow/client.ts:20]: duplicated

File: packages/react-auth/src/shared/session.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+export const getSessionData = (storageAPI = localStorage) => {
+  const data = storageAPI.getItem(SESSION_DATA_KEY);
+  const session_json = data ? JSON.parse(data) : {};
+
+  return session_json;
+};
+
+const mapTokenData = (data: any) => {
+  const tokenData: Token = (({
```
Comment (cj-f) [inline packages/react-auth/src/shared/session.ts:46]: this seems a little strange, just trying to remove email and profile, with a shallow copy you could just delete them explicitly

File: packages/react-auth/src/AuthProvider/index.tsx [hunk 4]
```diff
@@ -246,7 +258,23 @@ export const AuthProvider: FC<AuthProviderProps> = ({
       setError(msg);
       setLoading(false);
     } else {
-      axios
+      const { success, response } = await auth.post(infoURL, { code: code });
```
Comment (cj-f) [inline packages/react-auth/src/AuthProvider/index.tsx:261]: should the client just have the client endpoints as methods, so one for client/userinfo

File: packages/react-auth/src/ComponentAuthProvider/index.tsx [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+const ComponentAuthProvider: FC<ComponentAuthProps> = ({
+  config,
+  children,
+}) => {
+  const [authenticated, setAuthenticated] = useState<boolean>(false);
+  const [loading, setLoading] = useState<boolean>(true);
+  const [error, setError] = useState<APIResponse>();
+  const [user, setUser] = useState<AuthUser>();
+
+  const auth = new AuthNClient(config);
```
Comment (cj-f) [inline packages/react-auth/src/ComponentAuthProvider/index.tsx:51]: should be memoized

File: packages/react-auth/src/ComponentAuthProvider/index.tsx [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+export interface ComponentAuthContextType {
+  authenticated: boolean;
+  loading: boolean;
+  error: APIResponse | undefined;
+  user?: AuthUser;
+  login: () => void;
+  logout: () => void;
+  config: AuthNConfig;
```
Comment (cj-f) [inline packages/react-auth/src/ComponentAuthProvider/index.tsx:27]: should both provider just make the memoized client object available to the users.

I assume the client will grow to include all client endpoints like the session endpoints, and users may want to be all to easily call something like session list if they write a component to show users everywhere they are logged in

File: packages/react-auth/src/AuthProvider/index.tsx [hunk 7]
```diff
@@ -142,6 +125,8 @@ export const AuthProvider: FC<AuthProviderProps> = ({
   const [error, setError] = useState<string>("");
   const [user, setUser] = useState<AuthUser>();
 
+  const auth = new AuthNClient(config);
```
Comment (cj-f) [inline packages/react-auth/src/AuthProvider/index.tsx:128]: should be memoized

File: packages/react-auth/src/ComponentAuthProvider/index.tsx [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+  // update session data on step
+  useEffect(() => {
+    const sessionData = getSessionData();
+    sessionData.user = user;
+    saveSessionData(sessionData);
+
+    // eslint-disable-next-line
+  }, [user]);
+
+  const getSessionData = (): SessionData => {
```
Comment (cj-f) [inline packages/react-auth/src/ComponentAuthProvider/index.tsx:75]: shared function?

File: packages/react-auth/src/ComponentAuthProvider/index.tsx [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+    return {};
+  };
+
+  const saveSessionData = (data: SessionData) => {
+    const dataString = JSON.stringify(data);
+    localStorage.setItem(SESSION_NAME, dataString);
+  };
+
+  const validate = async (data: SessionData) => {
```
Comment (cj-f) [inline packages/react-auth/src/ComponentAuthProvider/index.tsx:91]: shared?

File: packages/react-auth/src/shared/utils.ts [hunk 10]
```diff
@@ -55,3 +55,22 @@ export const toUrlEncoded = (obj: { [key: string]: string }): string => {
...
+ * @returns {boolean}
+ * @example
+ * isLocalhost("pangea.cloud");
+ * // false
+ *
+ * isLocalhost("127.0.0.1");
+ * // true
+ */
+export const isLocalhost = (hostname: string): boolean => {
+  if (hostname === "localhost" || hostname === "127.0.0.1") {
```
Comment (r3dcrosse) [inline packages/react-auth/src/shared/utils.ts:71]: I know I was the one that wrote this, but wondering if it's worth adding a check for `0.0.0.0` here too?

File: packages/react-auth/src/shared/session.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    scopes,
+  }) => ({ id, identity, token, type, life, expire, created_at, scopes }))(
+    data
+  );
+  return tokenData;
+};
+
+export const getUserFromResponse = (data: APIResponse): AuthUser => {
+  const activeToken = { ...data.result.active_token };
+  const refreshToken = { ...data.result.refresh_token };
```
Comment (r3dcrosse) [inline packages/react-auth/src/shared/session.ts:63]: Do we anticipate `data.result` could ever be `undefined` at any point, like if the network request fails or something?

File: packages/react-auth/src/AuthProvider/index.tsx (hunk 4 above)
Comment (ggallien13) [reply packages/react-auth/src/AuthProvider/index.tsx]: yes, definitely should be a method. added a `userinfo` method on the client

File: packages/react-auth/src/AuthNClient/index.ts (hunk 1 above)
Comment (ggallien13) [reply packages/react-auth/src/AuthNClient/index.ts]: removed the "?"

File: packages/react-auth/src/AuthNFlow/client.ts (hunk 2 above)
Comment (ggallien13) [reply packages/react-auth/src/AuthNFlow/client.ts]: cleaned up

File: packages/react-auth/src/shared/session.ts (hunk 3 above)
Comment (ggallien13) [reply packages/react-auth/src/shared/session.ts]: yea, good suggestion... at first i was only going to copy a couple values. removed the copy, deleting two keys

File: packages/react-auth/src/ComponentAuthProvider/index.tsx (hunk 5 above)
Comment (ggallien13) [reply packages/react-auth/src/ComponentAuthProvider/index.tsx]: done

File: packages/react-auth/src/ComponentAuthProvider/index.tsx (hunk 8 above)
Comment (ggallien13) [reply packages/react-auth/src/ComponentAuthProvider/index.tsx:74]: yep, cleaned up

File: packages/react-auth/src/shared/utils.ts (hunk 10 above)
Comment (ggallien13) [reply packages/react-auth/src/shared/utils.ts:71]: i think `localhost` and `127.0.0.1` is good

File: packages/react-auth/src/shared/session.ts (hunk 11 above)
Comment (ggallien13) [reply packages/react-auth/src/shared/session.ts:47]: Currently, this function is only called before checking if `result.active_token.token` exists, and wouldn't be called on a response that didn't return a "Success" status, so there's always a result.

File: packages/react-auth/src/ComponentAuthProvider/index.tsx [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+  );
+
+  return (
+    <PangeaContext.Provider value={memoData}>
+      <>{children}</>
+    </PangeaContext.Provider>
+  );
+};
+
+export const usePangea = () => {
```
[the remark this answers is not in the document]
Comment (ggallien13) [reply packages/react-auth/src/ComponentAuthProvider/index.tsx]: yep, updated to be `useAuth`

File: packages/react-auth/src/AuthProvider/index.tsx (hunk 7 above)
Comment (ggallien13) [reply packages/react-auth/src/AuthProvider/index.tsx]: done

File: packages/react-auth/src/ComponentAuthProvider/index.tsx (hunk 9 above)
Comment (ggallien13) [reply packages/react-auth/src/ComponentAuthProvider/index.tsx:90]: The `validate` methods in the providers are different in how they handle the response, so I left them separate. Both call the same `client.validate()` method though.

**Pull request merged**